File: src/condition/Cron.php
<?php


namespace Ething\Condition;


	
class Cron extends Condition {
	
	
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 
				'cron' => null
			),
			$attributes
		);
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'cron':
					
					if(!is_string($attributes[$key]))
						throw new \Exception("field '{$key}' must be a string.");
					
					if(empty($attributes[$key]) || !\Cron\CronExpression::isValidExpression($attributes[$key]))
						throw new \Ething\Exception('invalid cron expression');
					
					break;
				
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true; 
	}
	
	protected function call(\Ething\Event\Signal $signal){
		$cron = \Cron\CronExpression::factory($this->cron);
		return $cron->isDue();
	}
	
}

File: ething/rule/condition/Cron.py
# coding: utf-8
from future.utils import string_types
from .. import Condition
from croniter import croniter
import time

DELTA = 2

class Cron(Condition):
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('cron', None)
        
        for key in attributes:
            
                
            if key == 'cron':
                
                expr = attributes[key]
                
                if not( isinstance(expr, string_types) and croniter.is_valid(expr) ):
                    raise Exception("not a valid cron expression.")
                
            else:
                raise Exception("%s: invalid" % key)
        
        return True
    
    
    
    def call(self, signal):
        
        ts = signal['ts']
        expr = self['cron']
        
        iter = croniter(expr, int(ts - DELTA))
        
        c = iter.get_current()
        n = iter.get_next()
        
        if n - c <= 2*DELTA:
            return True
        else:
            return False
    